src/core/meshcop/dtls.cpp

```diff
@@ -165,35 +219,137 @@ otError Dtls::Start(bool             aClient,
+
     VerifyOrExit(rval == 0);
 
-    mStarted = true;
+    mStarted                = true;
+    mConnectionClosedByHost = false;
     Process();
 
-    otLogInfoMeshCoP(GetInstance(), "DTLS started");
+    if (!mApplicationCoapSecure)
+    {
+        otLogInfoMeshCoP(GetInstance(), "DTLS started");
+    }
+    else
+    {
+        otLogInfoCoap(GetInstance(), "Application Coap Secure DTLS started");
+    }
+
+exit:
+    return MapError(rval);
+}
+
+#if OPENTHREAD_ENABLE_APPLICATION_COAP_SECURE
+
+otError Dtls::SetApplicationCoapSecureKeys(int *aCipherSuite, int aAnsCipherSuite)
+{
+    int rval = 0, i = 0;
+
+    for (i = 0; i < aAnsCipherSuite; i++)
+    {
+        VerifyOrExit(&aCipherSuite[i] != NULL, rval = MBEDTLS_ERR_SSL_BAD_INPUT_DATA);
+
+        switch (mApplicationCoapCiphreSuite[i])
+        {
+#ifdef MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
+
+        case MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8:
+
+            if (mCaChainSrc != NULL)
+            {
+                rval = mbedtls_x509_crt_parse(&mCaChain, (const unsigned char *)mCaChainSrc, (size_t)mCaChainLength);
+                VerifyOrExit(rval == 0);
+                mbedtls_ssl_conf_ca_chain(&mConf, &mCaChain, NULL);
+            }
+
+            if (mOwnCertSrc != NULL && mPrivateKeySrc != NULL)
+            {
+                rval = mbedtls_x509_crt_parse(&mOwnCert, (const unsigned char *)mOwnCertSrc, (size_t)mOwnCertLength);
+                VerifyOrExit(rval == 0);
+                rval = mbedtls_pk_parse_key(&mPrivateKey, (const unsigned char *)mPrivateKeySrc,
+                                            (size_t)mPrivateKeyLength, NULL, 0);
+                VerifyOrExit(rval == 0);
+                rval = mbedtls_ssl_conf_own_cert(&mConf, &mOwnCert, &mPrivateKey);
+                VerifyOrExit(rval == 0);
+            }
+
+            break;
+
+#endif // MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
+
+#ifdef MBEDTLS_KEY_EXCHANGE_PSK_ENABLED
+
+        case MBEDTLS_TLS_PSK_WITH_AES_128_CCM_8:
+
+            rval = mbedtls_ssl_conf_psk(&mConf, (unsigned char *)mPreSharedKey, mPreSharedKeyLength,
+                                        (unsigned char *)mPreSharedKeyIdentity, mPreSharedKeyIdLength);
+            VerifyOrExit(rval == 0);
+
+            break;
+
+#endif // MBEDTLS_KEY_EXCHANGE_PSK_ENABLED
+
+        default:
+            otLogCritCoap(this, "Application Coap Secure DTLS: Not supported cipher.");
+            rval = MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+            ExitNow();
+            break;
+        }
+    }
 
 exit:
     return MapError(rval);
 }
 
+void Dtls::SetSslAuthMode(bool aVerifyPeerCertificate)
+{
+    mVerifyPeerCertificate = aVerifyPeerCertificate;
+}
+
+#endif // OPENTHREAD_ENABLE_APPLICATION_COAP_SECURE
+
 otError Dtls::Stop(void)
 {
     mbedtls_ssl_close_notify(&mSsl);
-    Close();
+    // wait for ssl close notify from peer.
+    mConnectionClosedByHost = true;
+    mWaitForCloseNotifyTimer.Start(kWaitForCloseNotifyMilli);
```

Comment:
do we really need wait close notify?

Reply:
When the connection immediately close, the peer's close notify receive by host after finished the dtls session. so the host interpret this as error, and send back an dtls error message. The peer also has closed the dtls session until now, receive this dtls error, and also send back an dtls error.
Other solutions:
* Bind delayed to socket after close 
* Send no DTLS Error, if no dtls session is open

Reply:
According to [mbedtls_ssl_close_notify()](https://tls.mbed.org/api/ssl_8h.html#ac2c1b17128ead2df3082e27b603deb4c) stated, the ssl context MUST not be used after this call. I think we should close it here, because we don't handle the possible close notify from peer.

For the issue you described, I agree we could add some delay for another connection, however since we already close the previous ssl context, I think it's safe to reuse `mTimer` instead of introducing another specific timer for this task. What do you think?

Reply:
Ok. I close the dtls session at the old place.
It should be safe to use the existing timer. So i use now `mTimer` to Delay another, new connection.